### ISSUE TITLE
Discussion: do we need to compress the CTC_loss when Reduce == False in Espnet2 ?

### ISSUE THREAD

**Issue by shincling** [espnet/espnet#3339]
Hi, @kamo-naoyuki and others, 
recently when I conducted the experiments of multi-spk ASR in Espnet2, some unexpected things happen: 
- The performance and learning curve are heavily dependent on the number of GPU used in training. 
Here are two experiments with the same configuration except the No. of GPU: 

Moldel | Setting| CER(dev/tt) | WER(dev/tt) | 
-- | -- | -- | --
npc +Transformer | 1gpu 32bs | 10.41/17.29 | 17.16/28.66
npc +Transformer | 4gpu 32bs | 6.74/12.69 | 9.22/17.33

After checking the learning curve, I found the CTC loss part are quite different, like this:
![image](https://user-images.githubusercontent.com/14304866/124716111-47c12380-df36-11eb-955d-3414461f4704.png)

Then, I checked the code about the CTC part used in multi-spk experiments, and I found the reason:

- In multi-spk ASR case, basically, we need to set the Reduce == Fasle in ctc_conf in config file to calculate the best permutation for speakers. 

In current CTC class in espent2, https://github.com/espnet/espnet/blob/98d5f82043157ed3a823bb765ec85c965a930aa9/espnet2/asr/ctc.py#L98-L102
the CTC loss will be reduced by the times of batch-size whtn Reduce == False, which actually changes the original range of loss_ctc, and also the given weight of loss_ctc_weight  in CTC/Attention architecture.

I checked the implementation about this part in Espnet1, 
https://github.com/espnet/espnet/blob/08feae5bb93fa8f6dcba66760c8617a4b5e39d70/espnet/nets/pytorch_backend/ctc.py#L45-L50
I feel the built-in CTC in torch also returns the original loss for each sample when Reduce is false 
https://github.com/espnet/espnet/blob/08feae5bb93fa8f6dcba66760c8617a4b5e39d70/espnet/nets/pytorch_backend/e2e_asr_mix.py#L218 .

Back to the former experiments, when the number of GPU >1, actually the batch-size in each GPU is smaller than that when n_GPU =1, which leads to larger CTC_loss in each GPU and the overall averaged CTC_loss.

So, in my view, maybe we should control this compression of CTC_loss to avoid the obvious performance gap with different number of GPU ( actually the different batch-size ).   
After simply removing this, I observed the CTC_loss becomes stable with different settings.

BTW, for now, there is no other place to set touch this part except the multi-spk ASR related tasks. 
If this should be fixed, we could fix it with two options in https://github.com/espnet/espnet/pull/3226:

1. Remove the compress in CTC class https://github.com/espnet/espnet/blob/98d5f82043157ed3a823bb765ec85c965a930aa9/espnet2/asr/ctc.py#L101-L102
2. Keep the original CTC, recover the CTC original range after calculating it from espnet_model.py
https://github.com/LiChenda/espnet/blob/develop/espnet2/asr/espnet_enh_asr_model.py#L378

which one do you prefer?  

Best, 
Jing

**Comment by kamo-naoyuki** [comment]
I checked the original implementation: 
https://github.com/espnet/warp-ctc/blob/0705437521a1302f38692fb684f4743f8a85d324/pytorch_binding/warpctc_pytorch/__init__.py#L36-L51
I thought `loss / size` is compatible with warpctc, but this is my misunderstanding, sorry.

Please remove `/ size` if you need.

**Comment by shincling** [comment]
Got it. Thanks Kamo.

**Comment by shincling** [comment]
BTW, @LiChenda @Emrys365 do you guys think maybe this causes the instability of converging when training with joint enh-asr model?
